slowquant/unitary_coupled_cluster/optimizers.py

```diff
@@ -0,0 +1,312 @@
+            param_names: Names of parameters, used to index R.
+            maxiter: Maximum number of iterations (sweeps).
+            tol: Convergence tolerence.
+            callback: Callback function, takes only x (parameters) as an argument.
+        """
+        self._callback = callback
+        self.max_iterations = maxiter
+        self.threshold = tol
+        self.max_fail = 3
+        self._R = R
+        self._param_names = param_names
+
+    def minimize(self, f: Callable[[list[float]], float | list[float]], x0: Sequence[float]) -> Result:
+        """Run minimization.
+
+        Args:
+            f: Function to be minimzed, can only take one argument.
+            x: Changable parameters of f.
+
+        Returns:
+            Minimization results.
+        """
+        f_best = float(10**20)
+        x = list(x0).copy()
+        x_best = x.copy()
+        fails = 0
+        res = Result()
+        for _ in range(self.max_iterations):
+            for i, par_name in enumerate(self._param_names):
+                e_vals = get_energy_evals(f, x, i, self._R[par_name])
+                f_reconstructed = partial(reconstructed_f, energy_vals=e_vals, R=self._R[par_name])
+                values = f_reconstructed(np.linspace(-np.pi, np.pi, int(1e4)))
+                theta = np.linspace(-np.pi, np.pi, int(1e4))[np.argmin(values)]
+                res = scipy.optimize.minimize(f_reconstructed, x0=[theta], method="BFGS", tol=1e-12)
+                x[i] = res.x[0]
+                while x[i] < np.pi:
+                    x[i] += 2 * np.pi
+                while x[i] > np.pi:
+                    x[i] -= 2 * np.pi
+            f_tmp = f(x)
+            if isinstance(f_tmp, list):
+                # State-averaged case
+                f_new = float(np.mean(f_tmp))
+            else:
+                # Single state case
+                f_new = f_tmp
+            if abs(f_best - f_new) < self.threshold:
+                f_best = f_new
+                x_best = x.copy()
+                break
+            if (f_new - f_best) > 0.0:
+                fails += 1
+            else:
+                f_best = f_new
+                x_best = x.copy()
+            if fails == self.max_fail:
+                break
+            if self._callback is not None:
+                self._callback(x)
+        res.x = np.array(x_best)
+        res.fun = f_best
+        return res
+
+
+def get_energy_evals(
+    f: Callable[[list[float]], float | list[float]], x: list[float], idx: int, R: int
+) -> list[float] | list[list[float]]:
+    """Evaluate the function in all points needed for the reconstrucing in Rotosolve.
+
+    Args:
+        f: Function to evaluate.
+        x: Parameters of f.
+        idx: Index of parameter to be changed.
+        R: Parameter to control how many points are needed.
+
+    Returns:
+        All needed function evaluations.
+    """
+    e_vals = []
+    x = x.copy()
```

Comment:
No x_shift here, compared to circuit_wavefunction.py? Is it absorbed in the equation?